cmd/clusterctl/client/tree/discovery_test.go

```diff
@@ -25,9 +25,43 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/test"
 )
 
+func clusterObjectsWithResourceSet() []client.Object {
+	namespace := "ns1"
+	clusterObjs := test.NewFakeCluster(namespace, "cluster1").
+		WithControlPlane(
+			test.NewFakeControlPlane("cp").
+				WithMachines(
+					test.NewFakeMachine("cp1"),
+				),
+		).
+		WithMachineDeployments(
+			test.NewFakeMachineDeployment("md1").
+				WithMachineSets(
+					test.NewFakeMachineSet("ms1").
+						WithMachines(
+							test.NewFakeMachine("m1"),
+							test.NewFakeMachine("m2"),
+						),
+				),
+		).
+		Objs()
+
+	var cluster *clusterv1.Cluster
+	for _, obj := range clusterObjs {
+		if obj.DeepCopyObject().GetObjectKind().GroupVersionKind().Kind == "Cluster" {
```

Comment:
nit: (non-blocking)
```suggestion
		if obj.GetObjectKind().GroupVersionKind() == clusterv1.GroupVersion.WithKind("Cluster") {
```
Since you are casting it to v1 Cluster object better to check the whole thing.

Also you don't need to do DeepCopy of the object.